### PR TITLE
Refactor FFN

### PR DESCRIPTION
**Pull request by ShangtongZhang**
- Support copy/assignment CTOR

- Add a new function `SetTrainingData` for async rl methods
This is mainly for compute gradients
```
    learningNetwork.SetTrainingData(state, target);
    arma::mat gradients;
    learningNetwork.Gradient(learningNetwork.Parameters(), 0, gradients);
```

- Expose a private function `ResetParameters` to public

### CODE REVIEW

File: src/mlpack/methods/ann/init_rules/gaussian_init.hpp [hunk 1]
```diff
@@ -76,10 +76,10 @@ class GaussianInitialization
 
  private:
   //! Mean of the gaussian.
-  const double mean;
+  double mean;
```
Comment (ShangtongZhang) [inline src/mlpack/methods/ann/init_rules/gaussian_init.hpp]: class with const member variable doesn't support assignment operator, so I just removed the const qualifier.

File: src/mlpack/methods/ann/ffn_impl.hpp [hunk 2]
```diff
@@ -379,6 +384,48 @@ void FFN<OutputLayerType, InitializationRuleType>::Serialize(
...
+  {
+    this->network.push_back(boost::apply_visitor(copyVisitor, network.network[i]));
+  }
+};
+
+template<typename OutputLayerType, typename InitializationRuleType>
+FFN<OutputLayerType, InitializationRuleType>&
+FFN<OutputLayerType, InitializationRuleType>::operator = (const FFN& network)
+{
+  // Replicate layers from source network
```
Comment (zoq) [inline src/mlpack/methods/ann/ffn_impl.hpp]: We should deconstruct the network first, to avoid memory leaks. Also maybe it's a good idea to use the  copy-and-swap idiom here, could also be useful if we provide a move constructor, what do you think?

File: src/mlpack/methods/ann/ffn.hpp [hunk 3]
```diff
@@ -177,6 +186,20 @@ class FFN
...
   arma::mat& Parameters() { return parameter; }
 
+  /**
+   * Make the network ready for the given training data.
+   * This function won't actually trigger training process.
+   *
+   * @param predictors Input training variables.
+   * @param responses Outputs results from input training variables.
+   */
+  void SetTrainingData(const arma::mat& predictors, const arma::mat& responses);
```
Comment (zoq) [inline src/mlpack/methods/ann/ffn.hpp]: Do you think instead of using ``SetTrainingData`` followed by ``Gradient``we could provide another ``Gradient`` function, that takes the ``predictors`` and ``responses``?

File: src/mlpack/methods/ann/ffn_impl.hpp (hunk 2 above)
Comment (ShangtongZhang) [inline src/mlpack/methods/ann/ffn_impl.hpp]: It's the first time I hear about copy-and-swap idiom. It's really an interesting and useful design! Also I will look into move CTOR.

File: src/mlpack/methods/ann/ffn.hpp (hunk 3 above)
Comment (ShangtongZhang) [inline src/mlpack/methods/ann/ffn.hpp]: That sounds good, more easy to use ^.^

**Comment by ShangtongZhang** [comment]
Hope it's ready to merge now

File: src/mlpack/methods/ann/ffn.hpp [hunk 4]
```diff
@@ -215,6 +217,15 @@ class FFN
   void Forward(arma::mat&& input);
 
   /**
+   * Prepare the network for the given training data.
+   * This function won't actually trigger training process.
+   *
+   * @param predictors Input training variables.
+   * @param responses Outputs results from input training variables.
+   */
+  void SetTrainingData(const arma::mat& predictors, const arma::mat& responses);
```
Comment (zoq) [inline src/mlpack/methods/ann/ffn.hpp]: Do we have to provide this function? I guess it's only used internally, I'd like to keep the function set minimal, so maybe we can delete the function.

File: src/mlpack/tests/feedforward_network_test.cpp [hunk 5]
```diff
@@ -410,4 +410,18 @@ BOOST_AUTO_TEST_CASE(DropConnectNetworkTest)
       (dataset, labels, dataset, labels, 2, 10, 50, 0.2);
 }
 
+/**
+ * Test miscellaneous things of FFN.
+ */
```
Comment (rcurtin) [inline src/mlpack/tests/feedforward_network_test.cpp]: Could you be a little more specific in this comment please? :)

File: src/mlpack/methods/ann/ffn.hpp (hunk 4 above)
Comment (ShangtongZhang) [inline src/mlpack/methods/ann/ffn.hpp]: It's a common part of `Gradient()` and `Train()`, I marked it as private. Without it we have to do copy-paste for some code snippet. I'm not sure which is preferred. If we decide to keep it, maybe we can change the name of this function. I don't think current name is so elegant but I can't come up with a better one..

File: src/mlpack/tests/feedforward_network_test.cpp (hunk 5 above)
Comment (ShangtongZhang) [inline src/mlpack/tests/feedforward_network_test.cpp]: Sure

File: src/mlpack/methods/ann/ffn.hpp (hunk 4 above)
Comment (zoq) [inline src/mlpack/methods/ann/ffn.hpp]: You are right, avoiding code duplication. I think the function name is misleading, it's not necessarily the training set, what you do you think about ``ResetData(...)``?

Comment (ShangtongZhang) [inline src/mlpack/methods/ann/ffn.hpp]: That looks good. I have changed the name to `ResetData`.

File: src/mlpack/methods/ann/ffn.hpp [hunk 6]
```diff
@@ -217,13 +217,13 @@ class FFN
...
 
   /**
-   * Prepare the network for the given training data.
+   * Prepare the network for the given data.
    * This function won't actually trigger training process.
    *
-   * @param predictors Input training variables.
-   * @param responses Outputs results from input training variables.
+   * @param predictors Input data variables.
+   * @param responses Outputs results from input datagit variables.
```
Comment (zoq) [inline src/mlpack/methods/ann/ffn.hpp]: Looks like a combination of ``data`` and ``git``.

File: src/mlpack/methods/ann/ffn.hpp [hunk 7]
```diff
@@ -219,6 +248,12 @@ class FFN
    */
   void ResetGradients(arma::mat& gradient);
 
+  /**
+   * Swap the content of this network with given network.
+   * @param network Desired source network.
```
Comment (zoq) [inline src/mlpack/methods/ann/ffn.hpp]: I know it is a pedantic request, but could you add newline between the description and the function parameters?

Comment (ShangtongZhang) [inline src/mlpack/methods/ann/ffn.hpp]: Oh, I didn't realize this before :)

**Comment by zoq** [comment]
Thanks for the great work!
